### PR TITLE
[GCI - Chinemerem]: Code Refactor

### PR DESCRIPTION
**Pull request by felixfavour**
refactored code to use maps instead of objects and added ES6 Support

### CODE REVIEW

File: dataservice.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+let categoriesObject = require('./my-app-react/src/data/categories.json')
```
Comment (f4ww4z) [inline dataservice.js]: Can't this be a `const` ? Also rename to `categoriesData` or `categoriesJson`

File: dataservice.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+let categoriesObject = require('./my-app-react/src/data/categories.json') 
+let productsObject = require('./my-app-react/src/data/products.json')
```
Comment (f4ww4z) [inline dataservice.js]: Can't this be a `const` ? Also rename to `productsData` or `produtsJson`

File: dataservice.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+let combinedProductMap = new Map();
```
Comment (f4ww4z) [inline dataservice.js]: Have you seen the task description step 8 and 9? Some lines end with `;` and some doesn't. Integrate eslint and fix eslint issues first.

File: dataservice.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+let productMap = new Map()
+let categoryMap = new Map()
+
+exports.getProducts = () => {
```
Comment (f4ww4z) [inline dataservice.js]: This method shouldn't be exported.

File: dataservice.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    for (productIndex in productsObject.products) {
```
Comment (f4ww4z) [inline dataservice.js]: Use `for .. of` loop instead, to make it easier to read.

File: dataservice.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        let product = productsObject.products[productIndex]
+        productMap.set(product.id, product)
+    }
+    return productMap
+}
+
+exports.getCategories = () => {
```
Comment (f4ww4z) [inline dataservice.js]: This method shouldn't be exported - only `getCombinedProductMap()` can.

File: dataservice.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    for (categoryIndex in categoriesObject.categories) {
```
Comment (f4ww4z) [inline dataservice.js]: Use `for .. of` loop instead, to make it easier to read.

File: dataservice.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        let category = categoriesObject[categoryIndex]
+        categoryMap.set(category.id, category.categoryName)
+    }
+    return categoryMap
+}
+
+let combineProductsWithCategories = () => {
+    let catKeys = categoryMap.keys()
```
Comment (f4ww4z) [inline dataservice.js]: Call the `getProducts()` and `getCategories()` methods here.

File: dataservice.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    
+    for (index in categoriesObject.categories) {
+        let categoryKey = catKeys.next().value
+        let productsArray = []
+        for(i=1; i < categoryMap.size; i++) {
+            let productValue = productEntries.next().value
+            productsArray.push(productValue)
+
+            if(productsArray.length == 3) {
+                combinedProductMap.set(categoryKey, productsArray)
```
Comment (f4ww4z) [inline dataservice.js]: Your logic is wrong. The `id` of the product should be the key, and the details of the product, including its `categoryName` should be the value. The format of the `combinedProductMap` should be:
```json
"P001": {
  "categoryId": "cty01",
  "name": "",
  "unit": "",
  "pricePerUnit": 0,
  "stocks": 0,
  "categoryName": "",
},
"P002": {
  "categoryId": "cty01",
  "name": "",
  "unit": "",
  "pricePerUnit": 0,
  "stocks": 0,
  "categoryName": "",
},
...
```

File: server.js [hunk 10]
```diff
@@ -14,35 +15,28 @@ app.get('/info', (req, res) => {
...
 app.get('/products/all', (req, res) => {
-    res.json(products)
+    let productsMap = dataService.getProducts()
+    let productsArray = []
+
+    let getValues = (value, key, productsMap) => {productsArray.push(value)}
+
+    productsMap.forEach(getValues)
+
+    res.json(productsArray)
```
Comment (f4ww4z) [inline server.js]: Should return a map of products, not just an array of product objects. See above json example.

File: server.js [hunk 11]
```diff
@@ -14,35 +15,28 @@ app.get('/info', (req, res) => {
...
-    for (productID in productsLocal) {
-        let product = productsLocal[productID];
-        if(urlCategoryID == product.categoryId) {
-            arrayOfProducts.push(product)
-        }
-    }
+    let combinedProductMap = dataService.getCombinedProductMap()
     
-    res.json(arrayOfProducts)
+    res.json(combinedProductMap.get(urlCategoryID))
```
Comment (f4ww4z) [inline server.js]: You need to filter the map `combinedProductMap` so only those with the correct category is shown.

File: server.js (hunk 10 above)
Comment (felixfavour) [reply server.js]: Can I restart the project? In a fresh repository...

Comment (felixfavour) [reply server.js]: How do I send a Map as a JSON response?
image
That is all my browser is showing when I send my Map as a JSON response.

Comment (felixfavour) [reply server.js]: It actually just send a map object as you suggested but it only `res.json()` takes only JS objects and array. All attempts to convert the map to objects have proved futile.

**Pull request closed without merging**